zio-http/src/main/scala/zio/http/Form.scala

```diff
@@ -0,0 +1,71 @@
+package zio.http
+
+import java.net.URLEncoder
+import java.nio.charset.Charset
+import java.net.URLDecoder
+import zio._
+
+/**
+ * Represents a form that can be either multipart or url encoded.
+ */
+sealed trait Form { self =>
+
+  def +(field: (String, String)): Form = appendField(field._1, field._2)
+
+  def appendField(name: String, value: String): Form =
+    self match {
+      case Form.URLEncodedForm(encoding, fields) =>
+        Form.URLEncodedForm(encoding, fields + (name -> value))
+    }
+
+  def contentType = self match {
+    case Form.URLEncodedForm(_, _) => "application/x-www-form-urlencoded"
+  }
+
+  def fields: Map[String, String]
+
+}
+
+object Form {
+
+  private val `UTF-8` = Charset.forName("UTF-8")
+
+  /**
+   * Represents a `application/x-www-form-urlencoded` form.
+   *
+   * `Request.form` will automatically detect and set the `Content-Type` header
+   * to `application/x-www-form-urlencoded` for this request.
+   *
+   * @param value
+   */
+  private[zio] final case class URLEncodedForm(private[zio] encoding: Charset, fields: Map[String, String])
+      extends Form {
+    def toEncodedString: String = encodeFields(encoding, fields).map { case (k, v) => s"$k=$v" }.mkString("&")
+  }
+
+  object URLEncodedForm {
+    def decode(encoded: String, encoding: Charset): Task[Form] = {
+      val fields = ZIO.foreach(encoded.split("&")) { pair =>
+        val array = pair.split("=")
+        if (array.length == 2)
+          ZIO.attempt((URLDecoder.decode(array(0), encoding), URLDecoder.decode(array(1), encoding)))
+        else
+          ZIO.fail(
+            new IllegalArgumentException(s"Invalid form field.  Expected: '{name}={value}'', Got '$pair' instead."),
+          )
+
+      }
+
+      fields.map(fields => URLEncodedForm(encoding, fields.toMap))
+    }
+  }
+
+  def makeUrlEncoded(encoding: Charset, fields: Seq[(String, String)]): Form = {
```

Comment:
The order of these two parameters should be flipped, because the `fields` parameter is more important and always required, whereas the other one is optional.

Also these are quite lengthy names. Probably just `make` would suffice.

Reply:
There will be a variant for multipart forms as well so we need a way to differentiate the two.  I am open to suggestions :)

Reply:
I think `.apply` should get the most "general" constructor, maybe with a little overloading.

Esoteric constructors should use the naming convention `.fromXYZ`.

Sound ok?

Reply:
Refactor includes the naming suggestions.